evm/vm/base.py

```diff
@@ -123,29 +122,63 @@ def execute_bytecode(self,
         )
 
     @abstractmethod
-    def make_receipt(self, transaction, computation, state):
+    def make_receipt(self, base_header, transaction, computation, state):
         """
-        Make receipt.
+        Generate the receipt resulting from applying the transaction.
+
+        :param base_header: the header of the block before the transaction was applied.
+        :param transaction: the transaction used to generate the receipt
+        :param computation: the result of running the transaction computation
+        :param state: the resulting state, after executing the computation
+
+        :return: receipt
+        """
+        raise NotImplementedError("Must be implemented by subclasses")
+
+    @abstractmethod
+    def validate_transaction_against_header(self, base_header, transaction):
+        """
+        Validate that the given transaction is valid to apply to the given header.
+
+        :param base_header: header before applying the transaction
+        :param transaction: the transaction to validate
+
+        :raises: ValidationError if the transaction is not valid to apply
         """
         raise NotImplementedError("Must be implemented by subclasses")
 
-    def apply_transaction(self, transaction):
+    def apply_transaction(self, header, transaction):
         """
         Apply the transaction to the vm in the current block.
+
+        :param transaction: to apply
+        :param header: header of the block before application
         """
+        self.validate_transaction_against_header(header, transaction)
         state_root, computation = self.state.apply_transaction(transaction)
-        receipt = self.make_receipt(transaction, computation, self.state)
-        # TODO: remove this mutation.
-        self.state.gas_used = receipt.gas_used
+        receipt = self.make_receipt(header, transaction, computation, self.state)
 
-        new_header = self.block.header.copy(
-            bloom=int(BloomFilter(self.block.header.bloom) | receipt.bloom),
+        new_header = header.copy(
+            bloom=int(BloomFilter(header.bloom) | receipt.bloom),
             gas_used=receipt.gas_used,
             state_root=state_root,
         )
 
         return new_header, receipt, computation
 
+    def _apply_all_transactions(self, transactions, base_header):
```

Comment:
How do you feel about sneaking in type definitions for new / changed functions? I do realize it reduces conformity for that file in question when only few defs are typed while others are not but on the other hand it helps to work towards our goal to have full coverage and eventually enable stricter type checks to disallow untyped defs altogether.

Reply:
:balance_scale: I'm a little on the fence. These APIs will continue to change a lot. I guess I'm slightly :-1:  at the moment, but when I feel like things start to settle, I'll be a lot more keen.